### PR TITLE
[Oracle] Oracle Hook - automatically set current_schema when defined in Connection

### LINKED ISSUE #18664: [Oracle] Oracle Hook - make it possible to define a schema in the connection parameters
### Description

Currently the oracle hook it not setting a CURRENT_SCHEMA after connecting with the Database.
In a lot of use cases we have production and test databases with seperate connections and database schemas e.g. TEST.Table1, PROD.Table1
"Hard-coding" the database schema in SQL Scripts is not elegant due to having different Airflow Instances for developing and Production.

An Option would be to store the database schema in a airflow Variable and getting it into the sql script with JINJA.
In Large SQL files with several tables it is not elegant either, because for every table a query to the metadatabase is made.

Why not using the Schema parameter in the Airflow Connections and executing 
`ALTER SESSION SET CURRENT_SCHEMA='SCHEMA'`
right after successfully connecting to the database?

An alternative would be to use option `Connection.current_schema ` of Library cx_Oracle.
https://cx-oracle.readthedocs.io/en/6.4.1/connection.html#Connection.current_schema

### Use case/motivation

It makes Query development much easier by storing environment attributes directly in the Airflow Connection.
You have full flexibility without touching your SQL Script.
It makes separation of Test and Production environments and connections possible.

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by mehmax**
Per Default, Oracle Hook was using "scheme" field of the Connection only when generating DSN uri.

This PR adds functionality to Oracle Hook.
-> When a schema is provided in the Connection Parameters (maintained in the UI), then this schema is set as current_schema after connecting to the Database

This allows SQLs to be less static.

Before:
`SELECT * FROM PROD_SCHEMA.MY_TABLE`
`SELECT * FROM TEST_SCHEMA.MY_TABLE`

Now:
`SELECT * FROM MY_TABLE`

closes: #18664
**^ Add meaningful description above**

### CODE REVIEW

File: airflow/providers/oracle/hooks/oracle.py [hunk 1]
```diff
@@ -136,6 +137,12 @@ def get_conn(self) -> 'OracleHook':
         if mod is not None:
             conn.module = mod
 
+        # if Connection.schema is defined, set schema after connecting successfully
+        # cannot be part of conn_config
+        # https://cx-oracle.readthedocs.io/en/latest/api_manual/connection.html?highlight=schema#Connection.current_schema
+        if schema is not None:
+            conn.current_schema = schema
```
Comment (uranusjr) [inline airflow/providers/oracle/hooks/oracle.py:163]: This `current_schema` does not seem to be used anywhere? Why is it needed? It seems to always be the same as `conn.schema`? (Because `schema = conn.schema` on line 80 and it's never modified.

Comment (uranusjr) [reply airflow/providers/oracle/hooks/oracle.py:163]: Oh, never mind, `conn` here is the database connection, not Airflow's `Connection` class. I was confused.

File: airflow/providers/oracle/hooks/oracle.py [hunk 2]
```diff
@@ -90,8 +91,8 @@ def get_conn(self) -> 'OracleHook':
                 dsn = conn.host
                 if conn.port is not None:
                     dsn += ":" + str(conn.port)
-                if service_name or conn.schema:
-                    dsn += "/" + (service_name or conn.schema)
+                if service_name:
+                    dsn += "/" + service_name
```
Comment (uranusjr) [inline airflow/providers/oracle/hooks/oracle.py:106]: Why is this change needed?

Comment (mehmax) [reply airflow/providers/oracle/hooks/oracle.py:106]: With this PR I generally fixed the meaning of conn.schema.
In my opinion it was never correct to append conn.schema to the dsn.

According to [cx_Oracle Documentation](https://cx-oracle.readthedocs.io/en/latest/user_guide/connection_handling.html#easy-connect-syntax-for-connection-strings) easy connect dsn is built like this:
`dbhost.example.com:1984/orclpdb1`
where `orclpdb1` is the service_name, not the schema.

Without a service_name, cx_Oracle is not able to connect.

Comment (uranusjr) [reply airflow/providers/oracle/hooks/oracle.py:106]: Sorry, I meant to ask why the _change_ was needed, not this entire block. This used to also consider `schema`, but does not anymore. Why not?

Comment (mehmax) [reply airflow/providers/oracle/hooks/oracle.py:106]: I removed it from the dsn because when people are using `schema` now, it could lead to a wrong dsn
eg. (`host:port/schema`). 
Correct is only `host:port/service_name` 

Do you think I should revert this change and the adapted tests?

Comment (uranusjr) [reply airflow/providers/oracle/hooks/oracle.py:106]: No, it's OK, but this does mean we're making a backward incompatible change and will need to bump the provider's major version. I don't remember if you need to do anything specific in code to call this out, but can you edit the PR's description to make it clear this will need a new major version bump? (So the release manager can notice it when compiling the release.)

Comment (mehmax) [reply airflow/providers/oracle/hooks/oracle.py:106]: Hi @uranusjr, yes you are right. I thought of another solution now to make it more of a softer change.
Instead of not considering schema anymore, I added a warning to the user that they should switch to using conn.extra.service_name instead. WDYT?

File: airflow/providers/oracle/hooks/oracle.py [hunk 3]
```diff
@@ -90,8 +91,19 @@ def get_conn(self) -> 'OracleHook':
...
+                elif conn.schema:
+                    self.log.warning(
+                        """You are using the parameter conn.schema to pass the Service Name
+                        of your Oracle Database. In this version of the Hook, conn.schema
+                        is additionally used to set the 'current_schema' attribute of the
+                        Oracle DB Session.
+                        Please use the correct parameter conn.extra.service_name instead to
+                        prevent future issues.
+                        More Info: https://github.com/apache/airflow/pull/19084"""
+                    )
```
Comment (uranusjr) [inline airflow/providers/oracle/hooks/oracle.py]: Can we make this a `DeprecationWarning`? There are multiple instances you can search as examples.

**Comment by uranusjr** [comment]
Sorry for missing this. This is close to ready IMO.

File: airflow/providers/oracle/hooks/oracle.py (hunk 3 above)
Comment (mehmax) [reply airflow/providers/oracle/hooks/oracle.py]: Sure, will look at it!
I also thought of shortening the text a little. Maybe to this:
```
Using conn.schema to pass the Oracle Service Name is deprecated.
Please use conn.extra.service_name instead.
More Info: https://github.com/apache/airflow/pull/19084
```
What do you think about this?

Comment (uranusjr) [reply airflow/providers/oracle/hooks/oracle.py]: Probably don’t need the More Info part; the first two sentences are good enough for me.

Comment (mehmax) [reply airflow/providers/oracle/hooks/oracle.py]: Hi @uranusjr, sorry it took so long.
I changed the warning to `DeprecationWarning` and simplified the message.

File: airflow/providers/oracle/hooks/oracle.py (hunk 2 above)
Comment (uranusjr) [reply airflow/providers/oracle/hooks/oracle.py:106]: That sounds more user-friendly. Go ahead and make the change please!

**Comment by uranusjr** [comment]
The linter is complaining.

**Comment by potiuk** [comment]
Yeah . I am really just about to release (slightly delayed) provider's release so it would be great to get that one ien @mehmax

**Comment by mehmax** [comment]
> 

Should be fixed now :)

**Comment by mehmax** [comment]
> The linter is complaining.

Should be fixed now. Sorry

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Just a docker-compose failure which we will fix shorlty.

**Comment by malthe** [comment]
@potiuk this seems to have broken something in our system – just letting you know for release purposes while I investigate.

**Comment by malthe** [comment]
This seems to have been fixed in https://github.com/apache/airflow/pull/21524 but I wonder if this was intentional in the sense that the `None` vs empty string situation was just resolved by chance.

**Comment by potiuk** [comment]
> This seems to have been fixed in #21524 but I wonder if this was intentional in the sense that the `None` vs empty string situation was just resolved by chance.

This was the reason why I "held" the RC2 version for Oracle provider - and yeah, I believe @mehmax found the same compatibility provlem and fixed it (included in RC3). Any more comments @malthe  about your error to verify it was the same issue?

**Comment by malthe** [comment]
@mehmax I had a `conn.schema` which was an empty string so I needed the #21524 behavior there as well. I can confirm that rc3 works for me.
